source/extensions/filters/http/json_to_metadata/filter.cc

```diff
@@ -0,0 +1,376 @@
+    no_body.inc();
+    return;
+  }
+
+  absl::StatusOr<Json::ObjectSharedPtr> result =
+      Json::Factory::loadFromStringNoThrow(body->toString());
+  if (!result.ok()) {
+    ENVOY_LOG(debug, result.status().message());
+    non_json.inc();
+    handleAllOnError(rules, reported_flag);
+    return;
+  }
+
+  Json::ObjectSharedPtr body_json = std::move(result.value());
+  StructMap struct_map;
+  for (const auto& rule : rules) {
+    const auto& keys = rule.keys();
+    Json::ObjectSharedPtr node = body_json;
+    bool on_missing = false;
+    for (unsigned long i = 0; i < keys.size() - 1; i++) {
+      absl::StatusOr<Json::ObjectSharedPtr> next_node_result = node->getObjectNoThrow(keys[i]);
+      if (!next_node_result.ok()) {
+        ENVOY_LOG(warn, result.status().message());
+        handleOnMissing(rule, struct_map);
+        on_missing = true;
+        break;
+      }
+      node = std::move(next_node_result.value());
+    }
+    if (on_missing) {
+      continue;
+    }
+    absl::Status result = handleOnPresent(std::move(node), keys.back(), rule, struct_map);
+    if (!result.ok()) {
+      ENVOY_LOG(warn, fmt::format("{} key: {}", result.message(), keys.back()));
+      handleOnMissing(rule, struct_map);
+    }
+  }
+  success.inc();
+
+  finalizeDynamicMetadata(*decoder_callbacks_, struct_map, reported_flag);
+}
+
+void Filter::processRequestBody() {
+  processBody(decoder_callbacks_->decodingBuffer(), config_->requestRules(), is_request_reported_,
+              config_->stats().rq_success_, config_->stats().rq_no_body_,
+              config_->stats().rq_invalid_json_body_);
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  ASSERT(config_->doRequest());
+  if (!config_->requestContentTypeAllowed(headers.getContentTypeValue())) {
+    is_request_reported_ = true;
+    config_->stats().rq_mismatched_content_type_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (end_stream) {
+    handleAllOnMissing(config_->requestRules(), is_request_reported_);
+    config_->stats().rq_no_body_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance&, bool end_stream) {
+  ASSERT(config_->doRequest());
+  if (is_request_reported_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (end_stream) {
+    if (!decoder_callbacks_->decodingBuffer() ||
+        decoder_callbacks_->decodingBuffer()->length() == 0) {
+      handleAllOnMissing(config_->requestRules(), is_request_reported_);
+      config_->stats().rq_no_body_.inc();
+      return Http::FilterDataStatus::Continue;
+    }
+    processRequestBody();
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (decoder_callbacks_->decodingBuffer() &&
+      decoder_callbacks_->decodingBuffer()->length() > config_->requestBufferLimitBytes()) {
```

Comment:
Why not use the `requestBufferLimitBytes` to set the decoder_callbacks_ buffer limit to be independent of the "default behavior" of when Envoy stops buffering ? Also it seems like we will fail open, should this be a local reply?

Reply:
It's a good point and I didn't notice we can set that.
By reading the comments, seems not recommended to shrink the limit but not a hard limitation.
I agree that we shouldn't re-invent the wheel. So probably just set the bytes in `decodeHeaders` or even constructor?

https://github.com/envoyproxy/envoy/blob/ce9c61fe760fafa6d3033fc6bc45a013d37a8b5e/envoy/http/filter.h#L682-L693

Reply:
Yea I think we should leverage setting if this filters requests a higher limit than is present. If another filter has requested a higher limit, we must decide whether we want this to fail and send a local reply, etc.